Egamma PFID based on Tensorflow DNNs

valsdav opened the pull request:
#### PR description:

This PR introduces the DNN based Egamma particle flow ID evaluation based on the Tensorflow internal support.
Two helper classes, ElectronDNNEstimator and PhotonDNNEstimator, are included to handle the loading and evaluation of the DNN models for GsfElectron and GEDPhotons candidates. 

Multiple models are evaluated for electrons and photons in different detector and energy regions. Batch evaluation is implemented to call each Tensorflow session only 1 time for event.  The evaluation of the DNN models is configurable from python config and txt files:  the list of variables to be used, their scaling parameters, the model files can be changed without changes to the code.  

The DNNs are evaluated before the electron and photons candidates pass through the PFEgammaFilters, where the DNN values are available in the candidates to be used for the selection. Thresholds configuration for the PFid have been added to the PFEGammaFilter producer.

In the PR, the DNN based PFid has been activated by default and default models files have been included for the testing. Final model file will be provided in a separate PR after the finalization on the CMSSW_12_0  samples.  

Presentations and documentation:
- Physics talk in PPD:  https://indico.cern.ch/event/1084340/contributions/4560857/attachments/2328176/3966582/PPDMeeting_14Oct2021.pdf
- Integration of PF EGM ID in CMSSW  https://indico.cern.ch/event/1081058/#4-integration-of-pf-egm-id-in
- EGM PF validation studies https://indico.cern.ch/event/1081058/#5-egm-pf-validation-studies

#### PR validation:

Standard PF validation:  https://akanugan.web.cern.ch/akanugan/PF_Validation_PFEGammaID_PR35403/
List of MINIAOD files and a few summary slides looking at the PF electrons & gedGsfElectrons in test vs reference are at:
https://cernbox.cern.ch/index.php/s/ki533KAwuWW1k3T

valsdav committed:
Adding DNN variables for PFID for electrons in the dataformats

valsdav committed:
Evaluation of the PFID DNN also for photons

valsdav committed:
Added switch to disable the DNN PFID loading in Photon and Electron producers

jainshilpi committed:
PF cluster isolations

Added Pfiso variables for Photons
Added PFiso electron variables for DNN

valsdav committed:
Moved PFcluster based isolation for electrons in GsfElectronAlgo

valsdav committed:
Added protection for HovEn and changed name of Photons variables

Changed eta limits for photons also

valsdav committed:
Added default models and configurations for Run3 PFid

Improved naming of parameters in PFEgammaFilters

valsdav committed:
Added initial values for PFEgammaFilters DNN thresholds

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35403/25508

- This PR adds an extra 868KB to repository

- Found files with invalid states:
  - RecoEgamma/ElectronIdentification/data/ElectronDNNScaler.txt:
    - Added: 7cd103bc776df806ee3eacdc7fb755c5b1fb4cd6
    - Deleted: 9f23cd25ba419dbe9c04509c105f4a1c75a96e0f
  - RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/ElectronDNNScaler.txt:
    - Added: 3c2b7ba0c4ecd0f2f17566e16bb5c703af4135bc
    - Modified: dc76c55fd675b2adf7d1374432e52551b3608a17, 3e385acdeb8471b6e596dfe4f80893ad8a065cdd, 9f23cd25ba419dbe9c04509c105f4a1c75a96e0f
    - Deleted: 3caa230f17308c4ec37cab48c03486163c860258
  - RecoEgamma/PhotonIdentification/data/Photon_PFID_dnn/PhotonDNNScaler.txt:
    - Added: dc76c55fd675b2adf7d1374432e52551b3608a17
    - Modified: 9f23cd25ba419dbe9c04509c105f4a1c75a96e0f, 87e26409fd81bb01bf50acba1777df2ae0940800
    - Deleted: 3caa230f17308c4ec37cab48c03486163c860258

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/PatCandidates/src/classes_def_objects.xml  modified in PR(s): #31399
  - File  RecoParticleFlow/PFProducer/interface/PFEGammaFilters.h  modified in PR(s): #35309
  - File  RecoParticleFlow/PFProducer/python/particleFlow_cff.py  modified in PR(s): #35309
  - File  RecoParticleFlow/PFProducer/src/PFEGammaFilters.cc  modified in PR(s): #35309

cmsbuild commented:
A new Pull Request was created by  @valsdav (Davide Valsecchi) for master.

It involves the following packages:

- DataFormats/EgammaCandidates (**reconstruction**)
- DataFormats/ParticleFlowCandidate (**reconstruction**)
- DataFormats/PatCandidates (**reconstruction**)
- RecoEgamma/EgammaElectronAlgos (**reconstruction**)
- RecoEgamma/EgammaElectronProducers (**reconstruction**)
- RecoEgamma/EgammaIsolationAlgos (**reconstruction**)
- RecoEgamma/EgammaPhotonProducers (**reconstruction**)
- RecoEgamma/ElectronIdentification (**reconstruction**)
- RecoEgamma/PhotonIdentification (**reconstruction**)
- RecoParticleFlow/PFProducer (**reconstruction**)


@jpata, @cmsbuild, @slava77 can you please review it and eventually sign? Thanks.
@mmarionncern, @Sam-Harper, @sobhatta, @afiqaize, @jainshilpi, @cbernet, @rovere, @lgray, @gouskos, @lecriste, @hatakeyamak, @gpetruc, @wrtabb, @varuns23, @seemasharmafnal, @ram1123 this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
it would be nice to squash a few minor commits in the history.
42 is a bit large to start.

slava77 commented:
Files in RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/ should be removed from the commit history and instead proposed as a PR to https://github.com/cms-data/RecoEgamma-ElectronIdentification

I notice that the model files are as small as 15 kB. Is the model really so small/simple or are these "dummy" files?

slava77 commented:
@cmsbuild please test

slava77 commented:
abort test

slava77 commented:
enable profiling

slava77 commented:
@cmsbuild please test

valsdav commented:
HI @slava77 

> it would be nice to squash a few minor commits in the history.
> 42 is a bit large to start.

Of course, this can be done. Should I wait the end of the tests? 

> Files in RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/ should be removed from the commit history and instead proposed as a PR to https://github.com/cms-data/RecoEgamma-ElectronIdentification

I see, we can move them there. 

> 
> I notice that the model files are as small as 15 kB. Is the model really so small/simple or are these "dummy" files?

They are so small because the models are quite simple (DNN with 128-64-64-32-16-1 nodes) and they have been saved in tensorflow graph format with the optimization performed by this nice tool from the CMS ML group: https://cms-ml.github.io/documentation/inference/tensorflow2.html#saving-your-model

slava77 commented:
> > it would be nice to squash a few minor commits in the history.
> > 42 is a bit large to start.
> 
> Of course, this can be done. Should I wait the end of the tests?

the build was already done, so the commits can be updated any time now.

cmsbuild commented:
-1

**Failed Tests**: UnitTests RelVals AddOn 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/summary.html
**COMMIT**: 035207b9f7333b02939ef45b6e84e43e1a454ffb
**CMSSW**: CMSSW_12_1_X_2021-09-24-1100/slc7_amd64_gcc900
**Additional Tests**: PROFILING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/35403/18898/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test testRecoMETMETProducers had ERRORS
---> test testTauEmbeddingProducers had ERRORS
</pre>

## RelVals

- [140.53](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/140.53_RunHI2011+RunHI2011+RECOHID11+HARVESTDHI)
```
----- Begin Fatal Exception 24-Sep-2021 17:43:19 CEST-----------------------
An exception of category 'InvalidGraphDef' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=GsfElectronProducer label='gedGsfElectronsTmp'
Exception Message:
error while loading graphDef from 'RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/lowpT/lowpT_modelDNN.pb': Not found: RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/lowpT/lowpT_modelDNN.pb; No such file or directory
----- End Fatal Exception -------------------------------------------------
```
- [5.1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/5.1_TTbar+TTbarFS+HARVESTFS)
```
----- Begin Fatal Exception 24-Sep-2021 17:43:19 CEST-----------------------
An exception of category 'InvalidGraphDef' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=GsfElectronProducer label='gedGsfElectronsTmp'
Exception Message:
error while loading graphDef from 'RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/lowpT/lowpT_modelDNN.pb': Not found: RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/lowpT/lowpT_modelDNN.pb; No such file or directory
----- End Fatal Exception -------------------------------------------------
```
- [135.4](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/135.4_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS)
```
----- Begin Fatal Exception 24-Sep-2021 17:43:19 CEST-----------------------
An exception of category 'InvalidGraphDef' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=GsfElectronProducer label='gedGsfElectronsTmp'
Exception Message:
error while loading graphDef from 'RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/lowpT/lowpT_modelDNN.pb': Not found: RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/lowpT/lowpT_modelDNN.pb; No such file or directory
----- End Fatal Exception -------------------------------------------------
```
<details>
<summary>Expand to see more relval errors ...</summary>

- [4.22](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/4.22_RunCosmics2011A+RunCosmics2011A+RECOCOSD+ALCACOSD+SKIMCOSD+HARVESTDC)
- [136.7611](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/136.7611_RunJetHT2016E_reminiaod+RunJetHT2016E_reminiaod+REMINIAOD_data2016_HIPM+HARVESTDR2_REMINIAOD_data2016_HIPM)
- [140.56](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/140.56_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18)
- [4.53](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/4.53_RunPhoton2012B+RunPhoton2012B+HLTD+RECODR1reHLT+HARVESTDR1reHLT)
- [136.731](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/136.731_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2)
- [136.793](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/136.793_RunDoubleEG2017C+RunDoubleEG2017C+HLTDR2_2017+RECODR2_2017reHLT_skimDoubleEG_Prompt+HARVEST2017)
- [136.874](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/136.874_RunEGamma2018C+RunEGamma2018C+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Offline_L1TEgDQM+HARVEST2018_L1TEgDQM)
- [1000.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT)
- [1001.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/1001.0_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVDSIPIXELCALRUN1+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5)
- [1306.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/1306.0_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15)
- [9.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/9.0_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST)
- [1330.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15_L1TMuDQM+HARVESTUP15_L1TMuDQM+NANOUP15)
- [312.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/312.0_Pyquen_ZeemumuJets_pt10_2760GeV_2021+Pyquen_ZeemumuJets_pt10_2760GeV_2021+DIGIHI2021MIX+RECOHI2021MIX+HARVESTHI2021PPRECO)
- [25.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/25.0_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT)
- [10042.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/10042.0_ZMM_13+2017+ZMM_13TeV_TuneCUETP8M1_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano)
- [10024.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/10024.0_TTbar_13+2017+TTbar_13TeV_TuneCUETP8M1_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano)
- [10824.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/10824.0_TTbar_13+2018+TTbar_13TeV_TuneCUETP8M1_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano)
- [11634.7](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/11634.7_TTbar_14TeV+2021_trackingMkFit+TTbar_14TeV_TuneCP5_GenSim+Digi+Reco+HARVEST)
- [11634.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/11634.0_TTbar_14TeV+2021+TTbar_14TeV_TuneCP5_GenSim+Digi+Reco+HARVEST+ALCA)
- [25202.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25+NANOUP15_PU25)
- [12434.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/12434.0_TTbar_14TeV+2023+TTbar_14TeV_TuneCP5_GenSim+Digi+Reco+HARVEST+ALCA)
- [11634.912](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/11634.912_TTbar_14TeV+2021_DD4hepDB+TTbar_14TeV_TuneCP5_GenSim+Digi+Reco+HARVEST+ALCA)
- [11634.911](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/11634.911_TTbar_14TeV+2021_DD4hep+TTbar_14TeV_TuneCP5_GenSim+Digi+Reco+HARVEST+ALCA)
- [10224.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/10224.0_TTbar_13+2017PU+TTbar_13TeV_TuneCUETP8M1_GenSim+DigiPU+RecoFakeHLTPU+HARVESTFakeHLTPU+Nano)
- [23234.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/23234.0_TTbar_14TeV+2026D49+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal)
- [34634.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/34634.0_TTbar_14TeV+2026D76+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal)
- [28234.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/28234.0_TTbar_14TeV+2026D60+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal)
- [250202.181](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/250202.181_TTbar_13UP18+TTbar_13UP18+PREMIXUP18_PU25+DIGIPRMXLOCALUP18_PU25+RECOPRMXUP18_PU25+HARVESTUP18_PU25)
- [34834.999](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/runTheMatrix-results/34834.999_TTbar_14TeV+2026D76PU_PMXS1S2PR+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+PREMIX_PremixHLBeamSpot14PU+DigiTriggerPU+RecoGlobalPU+HARVESTGlobalPU)
</details>



## AddOn Tests

- [fastsim](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/addOnTests/fastsim)
```
----- Begin Fatal Exception 24-Sep-2021 17:43:14 CEST-----------------------
An exception of category 'InvalidGraphDef' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=GsfElectronProducer label='gedGsfElectronsTmp'
Exception Message:
error while loading graphDef from 'RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/lowpT/lowpT_modelDNN.pb': Not found: RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/lowpT/lowpT_modelDNN.pb; No such file or directory
----- End Fatal Exception -------------------------------------------------
```
- [fastsim1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/addOnTests/fastsim1)
```
----- Begin Fatal Exception 24-Sep-2021 17:43:14 CEST-----------------------
An exception of category 'InvalidGraphDef' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=GsfElectronProducer label='gedGsfElectronsTmp'
Exception Message:
error while loading graphDef from 'RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/lowpT/lowpT_modelDNN.pb': Not found: RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/lowpT/lowpT_modelDNN.pb; No such file or directory
----- End Fatal Exception -------------------------------------------------
```
- [fastsim2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/addOnTests/fastsim2)
```
----- Begin Fatal Exception 24-Sep-2021 17:43:14 CEST-----------------------
An exception of category 'InvalidGraphDef' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=GsfElectronProducer label='gedGsfElectronsTmp'
Exception Message:
error while loading graphDef from 'RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/lowpT/lowpT_modelDNN.pb': Not found: RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/lowpT/lowpT_modelDNN.pb; No such file or directory
----- End Fatal Exception -------------------------------------------------
```
<details>
<summary>Expand to see more addon errors ...</summary>

- [hlt_mc_Fake](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/addOnTests/hlt_mc_Fake)
- [hlt_mc_Fake1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/addOnTests/hlt_mc_Fake1)
- [hlt_mc_Fake2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/addOnTests/hlt_mc_Fake2)
- [hlt_mc_GRun](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/addOnTests/hlt_mc_GRun)
- [hlt_mc_HIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/addOnTests/hlt_mc_HIon)
- [hlt_mc_PIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/addOnTests/hlt_mc_PIon)
- [hlt_mc_PRef](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/addOnTests/hlt_mc_PRef)
- [hlt_data_Fake](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/addOnTests/hlt_data_Fake)
- [hlt_data_Fake1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/addOnTests/hlt_data_Fake1)
- [hlt_data_Fake2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/addOnTests/hlt_data_Fake2)
- [hlt_data_GRun](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/addOnTests/hlt_data_GRun)
- [hlt_data_HIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/addOnTests/hlt_data_HIon)
- [hlt_data_PIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/addOnTests/hlt_data_PIon)
- [hlt_data_PRef](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18898/addOnTests/hlt_data_PRef)
</details>

slava77 commented on RecoEgamma/ElectronIdentification/src/ElectronDNNEstimator.cc:
```diff
@@ -0,0 +1,275 @@
+  tensorflow::setLogging(cfg_.log_level);
+  // load the graph definition
+  LogDebug("EleDNNPFid") << "Loading " << nModels_ << " graphs";
+  for (const auto& model_file : cfg_.models_files) {
```
I think that these need to be  FileInPath instead of strings; apparently the tests are not running from /src directory.

hatakeyamak commented:
Just tagging @akanugan here who is running some check of this PR in addition to jenkins.
BTW, he is seeing the same error as reported above from jenkins..

valsdav committed:
edm::FileInPath for model and scalers files

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35403/25526

- This PR adds an extra 448KB to repository

- Found files with invalid states:
  - RecoEgamma/ElectronIdentification/data/ElectronDNNScaler.txt:
    - Added: 7cd103bc776df806ee3eacdc7fb755c5b1fb4cd6
    - Deleted: 9f23cd25ba419dbe9c04509c105f4a1c75a96e0f
  - RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/ElectronDNNScaler.txt:
    - Added: 3c2b7ba0c4ecd0f2f17566e16bb5c703af4135bc
    - Modified: dc76c55fd675b2adf7d1374432e52551b3608a17, 3e385acdeb8471b6e596dfe4f80893ad8a065cdd, 9f23cd25ba419dbe9c04509c105f4a1c75a96e0f
    - Deleted: 3caa230f17308c4ec37cab48c03486163c860258
  - RecoEgamma/PhotonIdentification/data/Photon_PFID_dnn/PhotonDNNScaler.txt:
    - Added: dc76c55fd675b2adf7d1374432e52551b3608a17
    - Modified: 9f23cd25ba419dbe9c04509c105f4a1c75a96e0f, 87e26409fd81bb01bf50acba1777df2ae0940800
    - Deleted: 3caa230f17308c4ec37cab48c03486163c860258

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/PatCandidates/src/classes_def_objects.xml  modified in PR(s): #31399
  - File  RecoParticleFlow/PFProducer/interface/PFEGammaFilters.h  modified in PR(s): #35309
  - File  RecoParticleFlow/PFProducer/python/particleFlow_cff.py  modified in PR(s): #35309
  - File  RecoParticleFlow/PFProducer/src/PFEGammaFilters.cc  modified in PR(s): #35309

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35403/25526/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35403/25526/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35403/25527

- This PR adds an extra 420KB to repository

- Found files with invalid states:
  - RecoEgamma/ElectronIdentification/data/ElectronDNNScaler.txt:
    - Added: 85c4d30f871a7d027f32311898ed9231e12acfe4
    - Deleted: a7c49aa0fe9ffbb66ff2544f76231952078cbc2e
  - RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/ElectronDNNScaler.txt:
    - Added: 85c4d30f871a7d027f32311898ed9231e12acfe4
    - Modified: afe9d7edda4f8e039464bf5e48ec6d79198a4420, c97bdc97547498b0af927af2b4d4d2373d597dd2, a7c49aa0fe9ffbb66ff2544f76231952078cbc2e
    - Deleted: 1bbb0787856adb8f8f79a5c4230efa52aac79b44
  - RecoEgamma/PhotonIdentification/data/Photon_PFID_dnn/PhotonDNNScaler.txt:
    - Added: afe9d7edda4f8e039464bf5e48ec6d79198a4420
    - Modified: a7c49aa0fe9ffbb66ff2544f76231952078cbc2e
    - Deleted: 1bbb0787856adb8f8f79a5c4230efa52aac79b44

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/PatCandidates/src/classes_def_objects.xml  modified in PR(s): #31399
  - File  RecoParticleFlow/PFProducer/interface/PFEGammaFilters.h  modified in PR(s): #35309
  - File  RecoParticleFlow/PFProducer/python/particleFlow_cff.py  modified in PR(s): #35309
  - File  RecoParticleFlow/PFProducer/src/PFEGammaFilters.cc  modified in PR(s): #35309

cmsbuild commented:
Pull request #35403 was updated. @jpata, @cmsbuild, @slava77 can you please check and sign again.

jpata commented:
@cmsbuild please test

cmsbuild commented:
-1

**Failed Tests**: RelVals 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18953/summary.html
**COMMIT**: 98f8fcdb58186a69b9c165244add0cfb4ad46027
**CMSSW**: CMSSW_12_1_X_2021-09-27-1100/slc7_amd64_gcc900
**Additional Tests**: PROFILING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/35403/18953/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals

- [4.22](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18953/runTheMatrix-results/4.22_RunCosmics2011A+RunCosmics2011A+RECOCOSD+ALCACOSD+SKIMCOSD+HARVESTDC)
```
----- Begin Fatal Exception 27-Sep-2021 17:15:32 CEST-----------------------
An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 160960 lumi: 277 event: 10003226 stream: 0
   [1] Running path 'dqmoffline_step'
   [2] Prefetching for module DQMMessageLogger/'DQMMessageLogger'
   [3] Prefetching for module LogErrorHarvester/'logErrorHarvester'
   [4] Prefetching for module PhotonIDProducer/'PhotonIDProd'
   [5] Calling method for module GEDPhotonProducer/'photons'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: std::vector<reco::PFCluster>
Looking for module label: particleFlowClusterECAL
Looking for productInstanceName: 

   Additional Info:
      [a] If you wish to continue processing events after a ProductNotFound exception,
add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.

----- End Fatal Exception -------------------------------------------------
```
- [136.7611](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/18953/runTheMatrix-results/136.7611_RunJetHT2016E_reminiaod+RunJetHT2016E_reminiaod+REMINIAOD_data2016_HIPM+HARVESTDR2_REMINIAOD_data2016_HIPM)
```
----- Begin Fatal Exception 27-Sep-2021 17:16:39 CEST-----------------------
An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 277069 lumi: 81 event: 36189035 stream: 0
   [1] Running path 'dqmofflineOnPAT_step'
   [2] Prefetching for module SingleTopTChannelLeptonDQM_miniAOD/'singleTopElectronMediumDQM_miniAOD'
   [3] Prefetching for module PATElectronSlimmer/'slimmedElectrons'
   [4] Prefetching for module PATElectronSelector/'selectedPatElectrons'
   [5] Prefetching for module PATElectronProducer/'patElectrons'
   [6] Prefetching for module ReducedEGProducer/'reducedEgamma'
   [7] Prefetching for module GEDPhotonProducer/'ootPhotons'
   [8] Calling method for module GEDPhotonProducer/'ootPhotonsTmp'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: std::vector<reco::PFCluster>
Looking for module label: particleFlowClusterECAL
Looking for productInstanceName: 

   Additional Info:
      [a] If you wish to continue processing events after a ProductNotFound exception,
add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.

----- End Fatal Exception -------------------------------------------------
```

jpata commented:
The test failure looks related to the changes in the PR, particleFlowClusterECAL is not found in certain cases.
@valsdav please check! :)

jpata commented on RecoEgamma/EgammaElectronProducers/plugins/GsfElectronProducer.cc:
```diff
@@ -22,19 +22,47 @@
+                     std::vector<tensorflow::Session*> tf_sessions,
+                     reco::VertexCollection const& vertices,
+                     bool dnnPFidEnabled) {
+    std::vector<GsfElectron::MvaOutput> mva_outputs;
```
reserve the size according to the electron collection

jpata commented on RecoEgamma/EgammaElectronProducers/plugins/GsfElectronProducer.cc:
```diff
@@ -22,19 +22,47 @@
+    if (dnnPFidEnabled) {
+      // Here send the list of electrons to the ElectronDNNEstimator and get back the values for all the electrons in one go
+      LogDebug("GsfElectronProducer") << "Getting DNN PFId for ele";
+      std::vector<std::array<float, 5>> dnn_ele_pfid = hoc->iElectronDNNEstimator->evaluate(electrons, tf_sessions);
```
```suggestion
      const auto& dnn_ele_pfid = hoc->iElectronDNNEstimator->evaluate(electrons, tf_sessions);
```

jpata commented on RecoEgamma/EgammaElectronProducers/plugins/GsfElectronProducer.cc:
```diff
@@ -22,19 +22,47 @@
+        iele++;
+        auto values = dnn_ele_pfid[iele];
+        // get the previous values
+        auto mvaOutput = mva_outputs[iele];
```
```suggestion
        const auto& mvaOutput = mva_outputs[iele];
```

jpata commented on RecoEgamma/EgammaElectronProducers/plugins/GsfElectronProducer.cc:
```diff
@@ -22,19 +22,47 @@
+      int iele = -1;
+      for (auto& el : electrons) {
+        iele++;
+        auto values = dnn_ele_pfid[iele];
```
```suggestion
        const auto& values = dnn_ele_pfid[iele];
```

jpata commented on RecoEgamma/EgammaElectronProducers/plugins/GsfElectronProducer.cc:
```diff
@@ -116,6 +152,9 @@ class GsfElectronProducer : public edm::stream::EDProducer<edm::GlobalCache<GsfE
   const bool resetMvaValuesUsingPFCandidates_;
+
+  bool dnnPFidEnabled_;
+  std::vector<tensorflow::Session*> elePFid_tfSessions;
```
When the TF session is owned by the producer, when running in multithreaded mode in production, each producer will load it's own copy of the models. This can quickly explode the memory usage.
It would be preferrable to load the sessions once in a global cache that is shared across all the producer copies in each thread.
See e.g. here: https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/RecoMET/METPUSubtraction/plugins/DeepMETProducer.cc#L15

valsdav replied:
In the latest commit I moved the TF:sessions inside the DNNEstimator classes, that are handled inside the global cache both for electrons and photons. 

I haven't been able to define them as std::atomic, thought. The problem is that I need to load a variable number of models and I cannot make atomic the vector of Session*/GraphDefs* pointers.  What do you suggest in this case? Thanks

jpata replied:
Sorry, I was not sufficiently precise here. If you check the other examples of Session and GraphDef in CMSSW, you should do as is done there. Session is owned by the EDProducer, the GraphDef by the Cache.

jpata commented on RecoEgamma/EgammaElectronProducers/plugins/GsfElectronProducer.cc:
```diff
@@ -249,6 +288,59 @@ void GsfElectronProducer::fillDescriptions(edm::ConfigurationDescriptions& descr
+    psd1.add<bool>("enabled", false);
+    psd1.add<std::string>("inputTensorName", "FirstLayer_input");
+    psd1.add<std::string>("outputTensorName", "sequential/FinalLayer/Softmax");
+    psd1.add<std::string>("logLevel", "2");  //0 debug, 1 info, 2 warning
```
1. it would be preferrable to rely on standard CMSSW EDM logging levels (LogWarning, LogInfo, LogDebug), so that things are consistent across different modules
2. if a custom logging solution is really required for EGamma (this should be discussed in e.g. core software IMO), please use at least ints for the level, not strings

valsdav replied:
The option is there only for the Tensorflow module internal logging.  We switched to integer (adding a to_string for the tensorflow::setLogging method constructor.  https://gitlab.cern.ch/mrieger/CMSSW-DNN#logging)

jpata commented on RecoEgamma/EgammaElectronProducers/plugins/GsfElectronProducer.cc:
```diff
@@ -415,6 +536,13 @@ GsfElectronProducer::GsfElectronProducer(const edm::ParameterSet& cfg, const Gsf
       consumesCollector());
+
+  // Config for PFID dnn
+  auto pset_dnn = cfg.getParameter<edm::ParameterSet>("EleDNNPFid");
```
```suggestion
  const auto& pset_dnn = cfg.getParameter<edm::ParameterSet>("EleDNNPFid");
```

jpata commented on RecoEgamma/EgammaElectronProducers/python/gedGsfElectrons_cfi.py:
```diff
@@ -19,9 +19,27 @@
     combinationRegressionWeightLabels = ["gedelectron_p4combination_offline"],
+
+    #Activate the evaluation of Egamma PFID DNN
+    EleDNNPFid= dict(
+        enabled = True,
+        inputTensorName = "FirstLayer_input",
+        outputTensorName = "sequential/FinalLayer/Softmax",
+        modelsFiles = [
+            "RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/lowpT/lowpT_modelDNN.pb",
+            "RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/highpTEB/highpTEB_modelDNN.pb",
+            "RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/highpTEE/highpTEE_modelDNN.pb"
+        ],
+        scalersFiles = [
+            "RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/lowpT/lowpT_scaler.txt",
+            "RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/highpTEB/highpTEB_scaler.txt",
+            "RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/highpTEE/highpTEE_scaler.txt"
+        ]
+    )    
```
it should not be necessary to duplicate here, as it was already specified in the fillDescriptions.

valsdav replied:
In my opinion it would be good to specify here again at least the models and scalers files.  As soon as the final version of the models will be available we will refer to them here with more specific names, different from the default ones.

jpata reviewed (commented):
Partial review comments up to and including gedGsfElectrons_cfi.py. More will follow later.

valsdav committed:
Disable PFiso computation if the DNN is disabled

Code improvements

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35403/25629

- This PR adds an extra 52KB to repository

- Found files with invalid states:
  - RecoEgamma/ElectronIdentification/data/ElectronDNNScaler.txt:
    - Added: 85c4d30f871a7d027f32311898ed9231e12acfe4
    - Deleted: a7c49aa0fe9ffbb66ff2544f76231952078cbc2e
  - RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/ElectronDNNScaler.txt:
    - Added: 85c4d30f871a7d027f32311898ed9231e12acfe4
    - Modified: afe9d7edda4f8e039464bf5e48ec6d79198a4420, c97bdc97547498b0af927af2b4d4d2373d597dd2, a7c49aa0fe9ffbb66ff2544f76231952078cbc2e
    - Deleted: 1bbb0787856adb8f8f79a5c4230efa52aac79b44
  - RecoEgamma/PhotonIdentification/data/Photon_PFID_dnn/PhotonDNNScaler.txt:
    - Added: afe9d7edda4f8e039464bf5e48ec6d79198a4420
    - Modified: a7c49aa0fe9ffbb66ff2544f76231952078cbc2e
    - Deleted: 1bbb0787856adb8f8f79a5c4230efa52aac79b44

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/PatCandidates/src/classes_def_objects.xml  modified in PR(s): #31399
  - File  RecoParticleFlow/PFProducer/interface/PFEGammaFilters.h  modified in PR(s): #35309
  - File  RecoParticleFlow/PFProducer/python/particleFlow_cff.py  modified in PR(s): #35309
  - File  RecoParticleFlow/PFProducer/src/PFEGammaFilters.cc  modified in PR(s): #35309

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35403/25629/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35403/25629/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

valsdav commented on DataFormats/EgammaCandidates/interface/GsfElectron.h:
```diff
@@ -706,7 +706,21 @@ namespace reco {
+      float dnn_e_bkgPhoton;
+      MvaOutput()
+          : status(-1),
+            mva_Isolated(-999999999.),
```
I think that the actual -9999 value does not have significance, but it is just a placeholder for the values of the MVA to distinguish between 0. and "not-computed".  I added a mvaNan constexpr variable.

valsdav committed:
Added new DNN info in the PFCandidate dataformat

valsdav committed:
Improvements for TF sessions and fixes

- Disabled missing PFCluster iso computation when DNN is not requested
- Added Era modifier to activate the PFid DNN only in Run3_common era
- Moved TF sessions in the DNNEstimator, so in the GlobalCache

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35403/25649

ERROR: Build errors found during clang-tidy run.
```
RecoParticleFlow/PFProducer/src/PFEGammaFilters.cc:481:1: error: version control conflict marker in file [clang-diagnostic-error]
<<<<<<< HEAD
^
Suppressed 825 warnings (821 in non-user code, 3 NOLINT, 1 with check filters).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

jpata commented:
Looks like something was committed unintentionally.
```
RecoParticleFlow/PFProducer/src/PFEGammaFilters.cc:481:1: error: version control conflict marker in file [clang-diagnostic-error]
<<<<<<< HEAD
```

valsdav commented:
> Looks like something was committed unintentionally.
> 
> ```
> RecoParticleFlow/PFProducer/src/PFEGammaFilters.cc:481:1: error: version control conflict marker in file [clang-diagnostic-error]
> <<<<<<< HEAD
> ```

I'm sorry but I don't understand this issue. I don't have that text in that file.. https://github.com/valsdav/cmssw/blob/egamma_tensorflow_pfid/RecoParticleFlow/PFProducer/src/PFEGammaFilters.cc#L481

smuzaffar commented:
@valsdav   please see the bottom of the PR. github interface suggests that it is not mergable ( there are already changes merged in couple of files which you have changed in this Pr ). please resolve the merge conflicits

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35403/25652

- This PR adds an extra 456KB to repository

- Found files with invalid states:
  - RecoEgamma/ElectronIdentification/data/ElectronDNNScaler.txt:
    - Added: ffbf08340bbfa2fd0a490a4d8fec968c00d59590
    - Deleted: e6d9c764f42b026de2d8ba18e8de6681ac7c08ba
  - RecoEgamma/ElectronIdentification/data/Ele_PFID_dnn/ElectronDNNScaler.txt:
    - Added: ffbf08340bbfa2fd0a490a4d8fec968c00d59590
    - Modified: a952dcf3313c0ebcffbb20ffd9e5fab39a319246, 9da80d941a339f4afaed52b992e50350fd4b9ca7, e6d9c764f42b026de2d8ba18e8de6681ac7c08ba
    - Deleted: 41fb1c778e0f8a1feae53c8b82ebe6b085f36daf
  - RecoEgamma/PhotonIdentification/data/Photon_PFID_dnn/PhotonDNNScaler.txt:
    - Added: a952dcf3313c0ebcffbb20ffd9e5fab39a319246
    - Modified: e6d9c764f42b026de2d8ba18e8de6681ac7c08ba
    - Deleted: 41fb1c778e0f8a1feae53c8b82ebe6b085f36daf

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/PatCandidates/src/classes_def_objects.xml  modified in PR(s): #31399

cmsbuild commented:
Pull request #35403 was updated. @jpata, @cmsbuild, @slava77 can you please check and sign again.

jpata commented:
enable profiling

jpata commented:
@cmsbuild please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/19289/summary.html
**COMMIT**: 075a55e55ab4c3c6c8cc8b679339c1c95403e394
**CMSSW**: CMSSW_12_1_X_2021-09-30-1100/slc7_amd64_gcc900
**Additional Tests**: PROFILING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/35403/19289/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 5098 differences found in the comparisons
 - DQMHistoTests: Total files compared: 40
 - DQMHistoTests: Total histograms compared: 3211080
 - DQMHistoTests: Total failures: 7554
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3203504
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 39 files compared)
 - Checked 169 log files, 37 edm output root files, 40 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_1_X_2021-09-30-1100+a0d08a/45871/triggerResults): no differences found

jpata commented on RecoEgamma/PhotonIdentification/src/PhotonDNNEstimator.cc:
```diff
@@ -0,0 +1,255 @@
+  std::vector<float> inputs;
+  // Loop on the list of requested variables and scaling values for the specific modelIndex
+  // Different type of scaling are available: 0=noscaling, 1=standard scaler, 2=minmax
+  for (auto& [varName, type, par1, par2] : featuresMap_[modelIndex]) {
```
I notice that the scaler is duplicated between the electron and photon estimator. Can this be put in a small class to reduce code duplication?

jpata commented on RecoEgamma/PhotonIdentification/src/PhotonDNNEstimator.cc:
```diff
@@ -0,0 +1,255 @@
+  LogDebug("PhotonDNNPFid") << "Working on " << photons.size() << " photons";
+
+  int iphoton = -1;
+  for (auto& photon : photons) {
```
```suggestion
  for (const auto& photon : photons) {
```

jpata commented on RecoEgamma/PhotonIdentification/src/PhotonDNNEstimator.cc:
```diff
@@ -0,0 +1,255 @@
+    LogDebug("PhotonDNNPFid") << "Run model: " << m << " with " << counts[m] << " photons";
+    tensorflow::run(sessions_[m], {{cfg_.inputTensorName, input_tensors[m]}}, {cfg_.outputTensorName}, &output);
+    // Get the output and save the PhotonDNNEstimator::nOutputs numbers along with the photon index
+    auto r = output[0].tensor<float, 2>();
```
```suggestion
    const auto& r = output[0].tensor<float, 2>();
```

jpata commented on RecoEgamma/PhotonIdentification/src/PhotonDNNEstimator.cc:
```diff
@@ -0,0 +1,255 @@
+      for (uint k = 0; k < PhotonDNNEstimator::nOutputs; k++)
+        result[k] = r(b, k);
+      // Get the original index of the electorn in the original order
+      int photon_index = photonIndexMap[m][b];
```
```suggestion
      const int photon_index = photonIndexMap[m][b];
```

jpata commented on RecoEgamma/ElectronIdentification/src/ElectronDNNEstimator.cc:
```diff
@@ -0,0 +1,285 @@
+
+    tensorflow::run(sessions_[m], {{cfg_.inputTensorName, input_tensors[m]}}, {cfg_.outputTensorName}, &output);
+    // Get the output and save the ElectronDNNEstimator::nOutputs numbers along with the ele index
+    auto r = output[0].tensor<float, 2>();
```
```suggestion
    const auto r = output[0].tensor<float, 2>();
```

jpata commented on RecoEgamma/ElectronIdentification/src/ElectronDNNEstimator.cc:
```diff
@@ -0,0 +1,285 @@
+  LogDebug("EleDNNPFid") << "Working on " << electrons.size() << " electrons";
+
+  int iele = -1;
+  for (auto& ele : electrons) {
```
```suggestion
  for (const auto& ele : electrons) {
```

jpata commented on RecoEgamma/PhotonIdentification/interface/PhotonDNNEstimator.h:
```diff
@@ -0,0 +1,67 @@
+#ifndef __RecoEgamma_PhotonIdentification_PhotonDNNEstimator_H__
```
```suggestion
#ifndef RecoEgamma_PhotonIdentification_PhotonDNNEstimator_h
```

jpata commented on RecoParticleFlow/PFProducer/src/PFEGammaFilters.cc:
```diff
@@ -92,26 +102,38 @@ bool PFEGammaFilters::passPhotonSelection(const reco::Photon& photon) const {
-      return false;
+  if (usePhotonPFidDNN_) {
+    // Run3 DNN based PFID
+    float dnn = photon.pfDNN();
```
```suggestion
    const auto dnn = photon.pfDNN();
```

jpata commented on RecoParticleFlow/PFProducer/src/PFEGammaFilters.cc:
```diff
@@ -92,26 +102,38 @@ bool PFEGammaFilters::passPhotonSelection(const reco::Photon& photon) const {
+  if (usePhotonPFidDNN_) {
+    // Run3 DNN based PFID
+    float dnn = photon.pfDNN();
+    double photEta = fabs(photon.eta());
```
```suggestion
    const auto photEta = std::abs(photon.eta());
```

jpata commented on RecoParticleFlow/PFProducer/src/PFEGammaFilters.cc:
```diff
@@ -92,26 +102,38 @@ bool PFEGammaFilters::passPhotonSelection(const reco::Photon& photon) const {
+    // Run3 DNN based PFID
+    float dnn = photon.pfDNN();
+    double photEta = fabs(photon.eta());
+    if (photEta <= 1.566) {
```
hardcoded number 1.566 to be read from a constexpr

jpata commented on RecoParticleFlow/PFProducer/src/PFEGammaFilters.cc:
```diff
@@ -137,30 +159,46 @@ bool PFEGammaFilters::passElectronSelection(const reco::GsfElectron& electron,
+  double eleEta = fabs(electron.eta());
+
+  if (useElePFidDNN_) {  // Use DNN for ele pfID >=CMSSW12_1
+    float dnn_sig = electron.dnn_signal_Isolated() + electron.dnn_signal_nonIsolated();
```
```suggestion
    const auto dnn_sig = electron.dnn_signal_Isolated() + electron.dnn_signal_nonIsolated();
```

jpata commented on RecoParticleFlow/PFProducer/src/PFEGammaFilters.cc:
```diff
@@ -137,30 +159,46 @@ bool PFEGammaFilters::passElectronSelection(const reco::GsfElectron& electron,
+  if (useElePFidDNN_) {  // Use DNN for ele pfID >=CMSSW12_1
+    float dnn_sig = electron.dnn_signal_Isolated() + electron.dnn_signal_nonIsolated();
+    if (electronPt > ele_iso_pt_) {
+      if (eleEta <= 1.566) {
```
constexpr

jpata commented on RecoParticleFlow/PFProducer/src/PFEGammaFilters.cc:
```diff
@@ -137,30 +159,46 @@ bool PFEGammaFilters::passElectronSelection(const reco::GsfElectron& electron,
+  } else {  // Use legacy MVA for ele pfID < CMSSW_12_1
+    if (electronPt > ele_iso_pt_) {
+      double isoDr03 = electron.dr03TkSumPt() + electron.dr03EcalRecHitSumEt() + electron.dr03HcalTowerSumEt();
+      if (eleEta <= 1.485 && isoDr03 < ele_iso_combIso_eb_) {
```
constexpr for 1.485

jpata commented on RecoParticleFlow/PFProducer/src/PFEGammaFilters.cc:
```diff
@@ -137,30 +159,46 @@ bool PFEGammaFilters::passElectronSelection(const reco::GsfElectron& electron,
-    } else if (eleEta > 1.485 && isoDr03 < ele_iso_combIso_ee_) {
-      if (electron.mva_Isolated() > ele_iso_mva_ee_)
-        passEleSelection = true;
+  double eleEta = fabs(electron.eta());
```
```suggestion
  const auto eleEta = std::abs(electron.eta());
```

jpata commented on RecoParticleFlow/PFProducer/src/PFEGammaFilters.cc:
```diff
@@ -137,30 +159,46 @@ bool PFEGammaFilters::passElectronSelection(const reco::GsfElectron& electron,
+      if (validHoverE || !badHcal_eleEnable_) {
         passEleSelection = true;
+      } else {
+        bool EE = (std::abs(electron.eta()) > 1.485);  // for prefer consistency with above than with E/gamma for now
```
constexpr

jpata commented on RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc:
```diff
@@ -46,13 +46,39 @@
+    // Here we will have to load the DNN PFID if present in the config
+    PhotonDNNEstimator::Configuration config;
+    auto pset_dnn = conf.getParameter<edm::ParameterSet>("PhotonDNNPFid");
+    bool dnnEnabled = pset_dnn.getParameter<bool>("enabled");
```
```suggestion
    const auto dnnEnabled = pset_dnn.getParameter<bool>("enabled");
```

jpata commented on RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc:
```diff
@@ -46,13 +46,39 @@
+  CacheData(const edm::ParameterSet& conf) {
+    // Here we will have to load the DNN PFID if present in the config
+    PhotonDNNEstimator::Configuration config;
+    auto pset_dnn = conf.getParameter<edm::ParameterSet>("PhotonDNNPFid");
```
```suggestion
    const auto& pset_dnn = conf.getParameter<edm::ParameterSet>("PhotonDNNPFid");
```

jpata commented on RecoEgamma/PhotonIdentification/interface/PhotonDNNEstimator.h:
```diff
@@ -0,0 +1,67 @@
+
+  const Configuration cfg_;
+
+  std::vector<tensorflow::GraphDef*> graphDefs_;  // --> Should be std::atomic but does not compile
```
can you clarify? what is the issue with atomic here?

jpata commented on RecoEgamma/ElectronIdentification/interface/ElectronDNNEstimator.h:
```diff
@@ -0,0 +1,68 @@
+#ifndef __RecoEgamma_ElectronIdentification_ElectronDNNEstimator_H__
```
```suggestion
#ifndef RecoEgamma_ElectronIdentification_ElectronDNNEstimator_h
```

jpata reviewed (commented):
I went through the rest of the changes, some major comments here:
- [x] the .pb and .txt files should be scrubbed from this PR with a rebase and moved to a cms-data PR in https://github.com/cms-data/RecoEgamma-ElectronIdentification, with a corresponding tag update in cmsdist
- [x] there is some duplication between PhotonDNNEstimator and ElectronDNNEstimator, it would be good to find a way to unify a bit. At least the scaler, graphdef loading/closing
- [x] thread safety of this code should be checked explicitly, as well as to make sure there is no memory leak (ensure the correct use of tensorflow::GraphDef and tensorflow::Session)
- [x] try to also check for all cases where you can use `const auto / auto&`. I tried to catch all, but probably some escaped my attention.

jpata commented:
- [ ] @mialiu149 @riga please check the TF usage in this PR and confirm this is OK for ML production.

valsdav committed:
Added switch to disable the DNN PFID loading in Photon and Electron producers

Added expections for missing values in the DNN estimators

Added switch to disable the DNN PFID loading in Photon and Electron producers

Added expections for missing values in the DNN estimators

valsdav committed:
Added minmax scaler to DNN estimator

valsdav committed:
Added new scaler option for Photon PFID DNN

Activate the PFID DNN by default

valsdav commented:
The data files have been removed and two PR have been created for the electrons and photons DNN models. 

- Ele: https://github.com/cms-data/RecoEgamma-ElectronIdentification/pull/23
- Photon: https://github.com/cms-data/RecoEgamma-PhotonIdentification/pull/9

We will proceed now with the code review, thanks for your comments.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35403/25700

- This PR adds an extra 16KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/PatCandidates/src/classes_def_objects.xml  modified in PR(s): #31399

cmsbuild commented:
Pull request #35403 was updated. @jpata, @cmsbuild, @slava77 can you please check and sign again.

akanugan commented:
Hi,
Here are the results from the Validation of this PR.

The standard validation plots are at: https://akanugan.web.cern.ch/akanugan/PF_Validation_PFEGammaID_PR35403/
List of MINIAOD files and a few summary slides looking at the PF electrons & gedGsfElectrons in test vs reference are at:
https://cernbox.cern.ch/index.php/s/ki533KAwuWW1k3T

jpata commented:
test parameters:
  - pull_requests = https://github.com/cms-data/RecoEgamma-ElectronIdentification/pull/23, https://github.com/cms-data/RecoEgamma-PhotonIdentification/pull/9
  - enable_tests = profiling

jpata commented:
@cmsbuild please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/19321/summary.html
**COMMIT**: 606b3c9a82c3894ac7d6c263c153adb90d54eefc
**CMSSW**: CMSSW_12_1_X_2021-09-30-2300/slc7_amd64_gcc900
**Additional Tests**: PROFILING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/35403/19321/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test run-pylint3 had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 5098 differences found in the comparisons
 - DQMHistoTests: Total files compared: 40
 - DQMHistoTests: Total histograms compared: 3211080
 - DQMHistoTests: Total failures: 7554
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3203504
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 39 files compared)
 - Checked 169 log files, 37 edm output root files, 40 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_1_X_2021-09-30-2300+a0d08a/45902/triggerResults): no differences found

jpata commented:
The error from pylint3 does not look related to this PR, also appears in https://github.com/cms-sw/cmssw/pull/35237.
```
===== Test "run-pylint3" ====
timeout: failed to run command 'pylint3': No such file or directory

---> test run-pylint3 had ERRORS
TestTime:0
^^^^ End Test run-pylint3 ^^^^
```

Profiling CPU budget:
```
baseline: 1.2      19.57  GsfElectronProducer::produce(edm::Event&, edm::EventSetup const&) [193]
this PR:  1.2      19.41  GsfElectronProducer::produce(edm::Event&, edm::EventSetup const&) [191]
```

Many small reco differences from PFCandidates onwards in downstream quantities (jets, MET, taus):
![image](https://user-images.githubusercontent.com/69717/135810568-a72c063d-5d7f-4fac-8161-695d6df7a23d.png)

jpata reviewed (commented):
To fix your issue with `std::atomic`, I think you need something like this, because copy or move of `std::atomic` is not allowed, but is required in the way you construct the `std::vector`. Here I implemented it for PhotonDNN, but the same applies for the ElectronDNN (note the issue below about duplication). Here's a stackoverflow [thread](https://stackoverflow.com/questions/13193484/how-to-declare-a-vector-of-atomic-in-c) with more details. 

```diff
diff --git a/RecoEgamma/PhotonIdentification/interface/PhotonDNNEstimator.h b/RecoEgamma/PhotonIdentification/interface/PhotonDNNEstimator.h
index 6be1d73467a..ac39c2d1fdc 100644
--- a/RecoEgamma/PhotonIdentification/interface/PhotonDNNEstimator.h
+++ b/RecoEgamma/PhotonIdentification/interface/PhotonDNNEstimator.h
@@ -49,7 +49,7 @@ private:
 
   const Configuration cfg_;
 
-  std::vector<tensorflow::GraphDef*> graphDefs_;  // --> Should be std::atomic but does not compile
+  std::vector<std::atomic<tensorflow::GraphDef*>> graphDefs_;  // --> Should be std::atomic but does not compile
   std::vector<tensorflow::Session*> sessions_;
 
   uint nModels_;
diff --git a/RecoEgamma/PhotonIdentification/src/PhotonDNNEstimator.cc b/RecoEgamma/PhotonIdentification/src/PhotonDNNEstimator.cc
index 88e95d9b609..b99c5235c90 100644
--- a/RecoEgamma/PhotonIdentification/src/PhotonDNNEstimator.cc
+++ b/RecoEgamma/PhotonIdentification/src/PhotonDNNEstimator.cc
@@ -25,7 +25,7 @@ PhotonDNNEstimator::PhotonDNNEstimator(std::vector<std::string>& models_files,
                             << " variables --> LOADED";
 }
 
-PhotonDNNEstimator::PhotonDNNEstimator(const Configuration& cfg) : cfg_(cfg) {
+PhotonDNNEstimator::PhotonDNNEstimator(const Configuration& cfg) : cfg_(cfg),graphDefs_(cfg_.models_files.size()) {
   // Init tensorflow sessions
   nModels_ = cfg_.models_files.size();
   initTensorFlowGraphs();
@@ -36,10 +36,10 @@ PhotonDNNEstimator::PhotonDNNEstimator(const Configuration& cfg) : cfg_(cfg) {
 }
 
 PhotonDNNEstimator::~PhotonDNNEstimator() {
-  for (auto g : graphDefs_)
-    if (g != nullptr)
-      delete g;
+  for (size_t i=0;i<graphDefs_.size();i++)
+    if (graphDefs_[i] != nullptr)
+      delete graphDefs_[i];
@@ -50,10 +50,10 @@ void PhotonDNNEstimator::initTensorFlowGraphs() {
   tensorflow::setLogging(std::to_string(cfg_.log_level));
   // load the graph definition
   LogDebug("PhotonDNNPFid") << "Loading " << nModels_ << " graphs";
+  size_t i=0;
   for (const auto& model_file : cfg_.models_files) {
-    tensorflow::GraphDef* graphDef =
-        tensorflow::loadGraphDef(edm::FileInPath(model_file).fullPath());  //-->should be atomic but does not compile
-    graphDefs_.push_back(graphDef);
+    graphDefs_[i] = tensorflow::loadGraphDef(edm::FileInPath(model_file).fullPath());
+    i++;
   }
   LogDebug("PhotonDNNPFid") << "Graphs loaded";
 }
```

- As stated in the inline comments, several issues were marked as resolved but not actually resolved. Please check.
- The duplication between the electron and photon estimators needs to be addressed.
- As stated inline, I believe the sharing of `tensorflow::Session` across threads in the cache is not correct. The sessions should be owned by the corresponding `EDProducer`, not the Cache object.

valsdav commented:
>     * As stated in the inline comments, several issues were marked as resolved but not actually resolved. Please check.
> 
>     * The duplication between the electron and photon estimators needs to be addressed.
> 

Hi Joosep, thanks for your review. I marked some issues as resolved but I haven't still committed the changes. I will push all my changes after I have addressed all the issues. 

>     * As stated inline, I believe the sharing of `tensorflow::Session` across threads in the cache is not correct. The sessions should be owned by the corresponding `EDProducer`, not the Cache object.

About the tensorflow::Sessions, probably I got confused by this previous comment: https://github.com/cms-sw/cmssw/pull/35403#discussion_r717555276 

Should I keep them in the producer as done previously?

Thanks a lot, I will be right back with fresh commits.

jpata commented:
>About the tensorflow::Sessions, probably I got confused by this previous comment: #35403 (comment)
> Should I keep them in the producer as done previously?

Right, I clarified my comment in https://github.com/cms-sw/cmssw/pull/35403#discussion_r721192068.

valsdav committed:
Code improvements

valsdav commented on RecoEgamma/ElectronIdentification/interface/ElectronDNNEstimator.h:
```diff
@@ -0,0 +1,68 @@
+  * 0 = Do not scale the variable
+  * 1 = standard norm. par1=mean, par2=std
+  * 2 = MinMax. par1=min, par2=max */
+  std::vector<std::vector<std::tuple<std::string, uint, float, float>>> featuresMap_;
```
Added a ScalerConfig structure in the new EgammaDNNHelper class.

valsdav committed:
Improved code for DNN estimator

The duplicated code has been moved in EgammaTools/interface/EgammaDNNHelper.h.
ElectronDNNEstimator and PhotonDNNestimator are now just used to create the correct list of variables and
provide a function to select the model to be used for each candidate.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35403/25749

- This PR adds an extra 340KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/PatCandidates/src/classes_def_objects.xml  modified in PR(s): #31399

cmsbuild commented:
Pull request #35403 was updated. @jpata, @cmsbuild, @slava77 can you please check and sign again.

jpata commented:
@cmsbuild please test

jpata reviewed (commented):
Can you confirm all comments were implemented? I still see some `//SJ` and `cout` that is in the commits.

I uploaded the profiles for the Run3 workflow (11834.21) from the previous bot run here:
- mem: https://cms-reco-profiling.web.cern.ch/cms-reco-profiling/cgi-bin/igprof-navigator/prs/35403/igprofMEM_step3
- cpu: https://cms-reco-profiling.web.cern.ch/cms-reco-profiling/cgi-bin/igprof-navigator/prs/35403/igprofCPU_step3

Can you please check the initialization and evaluation cost in memory & CPU (search for ElectronDNNEstimator, PhotonDNNEstimator and the relevant producers)? Is the Photon DNN actually enabled?

valsdav committed:
Removed logLevel from EgammaDNNHelper

valsdav committed:
Fixed bug in storing of DNN outputs in GsfElectronProducer

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a0d08a/19397/summary.html
**COMMIT**: c826912f40079f5ea6581b50969d257ac0fa3b41
**CMSSW**: CMSSW_12_1_X_2021-10-04-2300/slc7_amd64_gcc900
**Additional Tests**: PROFILING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/35403/19397/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 5012 differences found in the comparisons
 - DQMHistoTests: Total files compared: 40
 - DQMHistoTests: Total histograms compared: 3219394
 - DQMHistoTests: Total failures: 5183
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3214189
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 39 files compared)
 - Checked 169 log files, 37 edm output root files, 40 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_1_X_2021-10-04-2300+a0d08a/45966/triggerResults): no differences found

valsdav reviewed (commented):
I checked again all the review comments and fixed a couple of missing corrections. 

I'm testing locally and I will soon push the updates.

valsdav committed:
Added getters in PFCandidate, fixed missing scaling mode in EgammaDNNHelper

valsdav committed:
Disabled by default DNN Egamma PFID for pre4

cmsbuild commented:
Pull request #35403 was updated. @jpata, @cmsbuild, @slava77 can you please check and sign again.

jpata commented:
test parameters:
  - pull_requests = https://github.com/cms-data/RecoEgamma-ElectronIdentification/pull/23, https://github.com/cms-data/RecoEgamma-PhotonIdentification/pull/9

jpata commented:
@cmsbuild please test

since the PFID is now disabled, I also disabled the profiling, we expect no reco change.

perrotta commented:
> The usage of new DNN-based PF ID is switched off in the [last commit](https://github.com/cms-sw/cmssw/pull/35403/commits/9538034a0e115263dbfe4d34820a2c60c9ff3bba), which means old MVA-based PF ID will be used by default. This is done because this PR is still aiming for pre4, and there is another major change (mkFit) going in pre4. It is not preferable that two major changes go in same pre-release, because in central validations using relVal samples, if something suspicious come up then it would be hard to track down if its due to mkFit or due to PF ID. Switching off new PF ID ensures no physics change from this PR.

Please notice that, according to the decision taken yesterday at the ORP meeting, we are going to build 12_1_0_pre4 today (basically now: we were just waiting for the results of the tests of the last night IB to start the build).
Therefore, according to that decision, this PR will **not** enter pre4 (unless any request to stop that build in order to integrate this PR will arrive after this afternoon PPD ccordination).
In view of integrating it in pre5 it would be probably wise to re-enable PFID here (unless you decide to do it in two steps): in any case, better to wait for the results of the tests just launched by Joosep, which can be used to verify how the switch was implemented.

swagata87 commented:
Thanks Andrea and Joosep. This plan is fine from E/gamma side. Indeed it doesn't make sense to rush for pre4, we can rather focus on test/validate this PR in the extra time we got before pre5.

slava77 commented:
related to https://github.com/cms-sw/cmssw/pull/35403#issuecomment-935685256
when we discussed this PR a couple of RECO meetings ago I recall that there was an option that the variant with the preliminary training files would not become a default, but the PR code would be reviewed (including runtime with enabled) and merged so that the PF group can do relevant checks.

@hatakeyamak  you were a part of that discussion, perhaps you can clarify if instead enabling with the preliminary training by default is OK.

[I can't seem to find the `@` tags for PF as a git team, perhaps it was not propagated yet]

jpata commented:
@riga  @mialiu149 @valsdav may we ask this PR as well to be discussed at the upcoming ML production meeting on Oct 11: https://indico.cern.ch/event/1080303/, especially in the context of https://github.com/cms-sw/cmssw/issues/35541.

valsdav committed:
Enabled back Egm PFid by default; endRun -> endStream for sessions closing

valsdav committed:
Fixed code-checks, added const to PFCandidate getters

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35403/25829

- This PR adds an extra 40KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/PatCandidates/src/classes_def_objects.xml  modified in PR(s): #31399

valsdav committed:
From atomic to unique_ptr for Tensorflow GraphDef

jpata commented:
@cmsbuild please test

vlimant commented on RecoEgamma/EgammaPhotonProducers/python/gedPhotons_cfi.py:
```diff
@@ -94,7 +98,50 @@
-    checkHcalStatus = cms.bool(True)
+    checkHcalStatus = cms.bool(True),
+    PhotonDNNPFid = cms.PSet(
+        enabled = cms.bool(False),
```
was this meant to be enabled ? ref #42949

swagata87 replied:
its enabled via era, as it is meant to work only for run 3.

https://github.com/cms-sw/cmssw/blob/aca7593194f3f57ea0b3ac0c82100d11b098d708/RecoEgamma/EgammaPhotonProducers/python/gedPhotonSequence_cff.py#L58-L62
